### PR TITLE
[gui] Add wareframe mode for mesh & mesh_instance, add slider_int for Window.GUI.

### PR DESCRIPTION
**Pull request by Morcki**
**Note:**
New features:
1. Support three display modes: `Fill Mode` (fill faces), `Line Mode` (Wareframe), `Point Mode` (rendering as points).
2. Add a new widget `slider_int` in `window.GUI`, which supports slide int type value from minimum to maximum.
3. Add an arg named `instance_count` to control the count of mesh instances.

**Usage:**
```python

window = ti.ui.Window("Display Mesh", (1024, 1024), vsync=True)
canvas = window.get_canvas()
scene = ti.ui.Scene()
camera = ti.ui.make_camera()

# slider_int usage
some_int_type_value = 0
def show_options():
    global some_int_type_value

    window.GUI.begin("Display Panel", 0.05, 0.1, 0.2, 0.15)
    display_mode = window.GUI.slider_int("Value Range", some_int_type_value, 0, 5)
    window.GUI.end()

while window.running:
    
    ...
    # if to show wareframe 
    scene.mesh_instance(vertices, indices, instance_count = 100 , show_wareframe = True)
    
    canvas.scene(scene)
    show_options()
    window.show()
```

### CODE REVIEW

File: python/taichi/ui/scene.py [hunk 1]
```diff
@@ -205,6 +206,9 @@ def mesh(self,
             index_count (int, optional):
                 only available when `indices` is provided, which is the the number
                 of vertices to draw.
+            display_mode (int, optional):
```
Comment (ailzhang) [inline python/taichi/ui/scene.py:209]: It's really error prone and hard for users to memorize these, try using an enum instead of ints?

File: python/taichi/ui/scene.py [hunk 2]
```diff
@@ -301,14 +312,17 @@ def mesh_instance(self,
         if transforms and (transforms.m != 4 or transforms.n != 4):
             print("Error! Transform matrix must be 4x4 shape")
             exit()
+        if display_mode < 0 or display_mode >= 3:
+            print("Error! display_mode must be 0(Fill), 1(Line), 2(Point)")
+            exit()
```
Comment (ailzhang) [inline python/taichi/ui/scene.py:317]: usually a better way to do checks is raising errors instead of explicit print + exit

File: taichi/python/export_ggui.cpp [hunk 3]
```diff
@@ -232,6 +238,7 @@ struct PyScene {
     renderable_info.draw_first_index = (int)draw_first_index;
     renderable_info.draw_vertex_count = (int)draw_vertex_count;
     renderable_info.draw_first_vertex = (int)draw_first_vertex;
+    renderable_info.display_mode = taichi::lang::PolygonMode(display_mode);
```
Comment (ailzhang) [inline taichi/python/export_ggui.cpp:241]: Yea this is really error prone(and why you should just expose PolygonMode enum to python instead) : imagine what if you inserted a new entry in the beginning of PolygonMode...

**Review by ailzhang — CHANGES_REQUESTED** [review]
Thanks! Had a small suggestion w.r.t using enums instead of raw ints

File: python/taichi/ui/scene.py (hunk 2 above)
Comment (Morcki) [reply python/taichi/ui/scene.py:317]: Sorry, I will fix that.

File: python/taichi/ui/scene.py (hunk 1 above)
Comment (Morcki) [reply python/taichi/ui/scene.py]: Yeah, I add a new class named DisplayMode to support passing enum numbers in newer updates.

File: python/taichi/ui/scene.py (hunk 2 above)
Comment (Morcki) [reply python/taichi/ui/scene.py]: Have changed to raise exception.

File: taichi/python/export_ggui.cpp (hunk 3 above)
Comment (Morcki) [reply taichi/python/export_ggui.cpp]: Fixed that.

File: taichi/ui/common/app_config.h [hunk 4]
```diff
@@ -3,6 +3,7 @@
 #include <string>
 #include "taichi/ui/utils/utils.h"
 #include "taichi/rhi/arch.h"
+#include "taichi/rhi/vulkan/vulkan_device.h"
```
Comment (AmesingFlank) [inline taichi/ui/common/app_config.h:6]: is this change needed?

File: taichi/ui/backends/vulkan/renderable.cpp [hunk 5]
```diff
@@ -48,6 +48,16 @@ void Renderable::update_data(const RenderableInfo &info) {
     prog->flush();
   }
 
+  bool is_pipeline_reset = false;
```
Comment (AmesingFlank) [inline taichi/ui/backends/vulkan/renderable.cpp:51]:
```suggestion
  bool needs_pipeline_reset = false;
```

File: taichi/ui/backends/vulkan/renderer.h [hunk 6]
```diff
@@ -78,7 +78,7 @@ class TI_DLL_EXPORT Renderer {
   taichi::lang::StreamSemaphore get_render_complete_semaphore();
 
  private:
-  glm::vec3 background_color_ = glm::vec3(0.f, 0.f, 0.f);
+  glm::vec3 background_color_ = glm::vec3(0.f, 0.f, 0.0f);
```
Comment (AmesingFlank) [inline taichi/ui/backends/vulkan/renderer.h:81]: why is this needed?

Comment (Morcki) [reply taichi/ui/backends/vulkan/renderer.h:81]: Sorry, I want to change `background_color_` to a default color, but I found that some tests without calling `clear_color` failed to pass. So later I cancled this feature. Just handwriting fault.

File: taichi/ui/common/app_config.h (hunk 4 above)
Comment (Morcki) [reply taichi/ui/common/app_config.h]: I will drop that change.

File: taichi/python/export_ggui.cpp [hunk 7]
```diff
@@ -232,13 +239,16 @@ struct PyScene {
...
     renderable_info.draw_first_vertex = (int)draw_first_vertex;
+    renderable_info.display_mode = display_mode;
 
     MeshInfo info;
     info.renderable_info = renderable_info;
     info.color = tuple_to_vec3(color);
     info.two_sided = two_sided;
     if (transforms.valid) {
-      info.num_instances = transforms.shape[0];
+      info.num_instances = draw_instance_count > transforms.shape[0]
```
Comment (YuCrazing) [inline taichi/python/export_ggui.cpp:249]: Maybe  `draw_instance_count > transforms.shape[0] - instance_offset` ? or something else.

File: taichi/python/export_ggui.cpp [hunk 8]
```diff
@@ -232,13 +239,16 @@ struct PyScene {
...
+                               ? transforms.shape[0]
+                               : (int)draw_instance_count;
```
Comment (YuCrazing) [inline taichi/python/export_ggui.cpp:251]: Maybe `transforms.shape[0] - draw_instance_count`.

**Review by YuCrazing — CHANGES_REQUESTED** [review]

File: python/taichi/ui/scene.py [hunk 9]
```diff
@@ -267,9 +294,11 @@ def mesh_instance(self,
             transforms (ti.Matrix.field, optional):
                 The Matrix must be 4x4 size with N instances, and data type should
                 be ti.f32, ti.i32, ti.u32. If None, then it behaves like raw mesh (no copy).
-            draw_first_instance (int, optional):
+            instance_offset (int, optional):
                 Default value is 0 which means no offset to show mesh instances. Otherwise,
-                the mesh instances will show from the `draw_first_instance`.
+                the mesh instances will show from the `instance_offset`.
+            instance_count (int, optional):
+                Default value is None which takes the shape value of transforms.
```
Comment (YuCrazing) [inline python/taichi/ui/scene.py:301]:
```suggestion
                The default value is None. If this parameter is not provided, instance_count = transforms.shape[0] - instance_offset.
```

**Review by YuCrazing — CHANGES_REQUESTED** [review]

File: taichi/python/export_ggui.cpp (hunk 7 above)
Comment (Morcki) [reply taichi/python/export_ggui.cpp]: Thanks, have solved this problem.

**Review by YuCrazing — APPROVED** [review]
LGTM!

File: taichi/ui/backends/vulkan/renderable.h [hunk 10]
```diff
@@ -40,6 +40,7 @@ struct RenderableConfig {
   std::string fragment_shader_path;
   taichi::lang::TopologyType topology_type{
       taichi::lang::TopologyType::Triangles};
+  taichi::lang::PolygonMode polygon_type{taichi::lang::PolygonMode::Fill};
```
Comment (ailzhang) [inline taichi/ui/backends/vulkan/renderable.h:43]:
```suggestion
  taichi::lang::PolygonMode polygon_mode{taichi::lang::PolygonMode::Fill};
```

File: taichi/python/export_ggui.cpp [hunk 11]
```diff
@@ -218,11 +223,13 @@ struct PyScene {
                      py::tuple color,
                      bool two_sided,
                      FieldInfo transforms,
+                     float draw_instance_count,
                      float draw_first_instance,
                      float draw_index_count,
```
Comment (ailzhang) [inline taichi/python/export_ggui.cpp:228]: nit: I somehow still prefer these counts to be of int type (esp. in C++) to avoid confusing developers, but we can discuss those separately and address that in a different PR if applicable

File: python/taichi/ui/scene.py [hunk 12]
```diff
@@ -13,6 +13,27 @@
 normals_field_cache = {}
 
 
+class DisplayMode_:
+    Fill = _ti_core.DisplayMode.Fill if _ti_core.GGUI_AVAILABLE else None
```
Comment (ailzhang) [inline python/taichi/ui/scene.py:17]: Your DisplayMode shouldn't be dependent on whether GGUI_AVAILABLE IMHO. The enum can be always available, whether it's usable is controlled by GGUI availability.

File: python/taichi/ui/scene.py [hunk 13]
```diff
@@ -13,6 +13,27 @@
...
+    Line = _ti_core.DisplayMode.Line if _ti_core.GGUI_AVAILABLE else None
+    Point = _ti_core.DisplayMode.Point if _ti_core.GGUI_AVAILABLE else None
+
+    @staticmethod
+    def __call__(mode: int):
```
Comment (ailzhang) [inline python/taichi/ui/scene.py:22]: This doesn't make sense hmmm: it seems like you never used this call method at all...

**Review by ailzhang — CHANGES_REQUESTED** [review]
Looks a lot better! I had a few more nits inline and then this PR is good to go

File: python/taichi/ui/scene.py (hunk 13 above)
Comment (Morcki) [reply python/taichi/ui/scene.py:22]: I would like to use display mode through an int value with imgui to control it every frame. As below:

```python
display_mode = 0
def show_options(): # You can choose what display mode you like in realtime
    global display_mode    
    window.GUI.begin("Display Panel", 0.05, 0.1, 0.2, 0.15)
    # display_mode (0:Fill Mode, 1:Line Mode, 2:Point Mode)
    display_mode = window.GUI.slider_int("display_mode", display_mode, 0, 2)
    window.GUI.end()

scene.mesh_instance(vertices, indices, instance_count = 100 ,display_mode=ti.ui.DisplayMode(display_mode)))
```
Here, I want to use `ti.ui.DisplayMode(display_mode)` instead of `if else` blablabla.

File: taichi/python/export_ggui.cpp (hunk 11 above)
Comment (Morcki) [reply taichi/python/export_ggui.cpp:228]: Ok, maybe we can talk about this deeply later.

File: taichi/ui/backends/vulkan/renderable.h (hunk 10 above)
Comment (Morcki) [reply taichi/ui/backends/vulkan/renderable.h:43]: Ok, I will replace that.

File: python/taichi/ui/scene.py (hunk 12 above)
Comment (Morcki) [reply python/taichi/ui/scene.py:17]: Still need it because every type in GGUI should check it's availability. And previously, I submitted the code without checking `GGUI_AVAILABLE` failed on some platforms.
ref:
1. https://github.com/taichi-dev/taichi/blob/7a8a9283535fbe9a17ee46132acd2fa4e7555c1f/python/taichi/ui/scene.py#L75-L80
2. https://github.com/taichi-dev/taichi/blob/7a8a9283535fbe9a17ee46132acd2fa4e7555c1f/python/taichi/ui/ui.py#L29
3. https://github.com/taichi-dev/taichi/blob/7a8a9283535fbe9a17ee46132acd2fa4e7555c1f/python/taichi/ui/ui.py#L12-L24

Comment (ailzhang) [reply python/taichi/ui/scene.py:17]: For the second case about ProjectionMode I think we'd better remove the GGUI_AVAILABLE check. 
Scattering this check around isn't the best practice and we only need to check it when we create a GGUI resource like scene or camera you mentioned above (which really depends on whether it's available), but these enums can be available no matter what. 
For example, in the code snippet you showed in the pr description, you should check GGUI availability when you create a window or scene, instead of when creating a displaymode enum.

File: python/taichi/ui/scene.py (hunk 13 above)
Comment (ailzhang) [reply python/taichi/ui/scene.py:22]: as discussed offline, a more user friendly implementation should show verbose enum name in strings instead of ints. e.g https://github.com/ocornut/imgui/blob/master/imgui_demo.cpp#L738-L747

File: python/taichi/ui/scene.py (hunk 12 above)
Comment (PENGUINLIONG) [reply python/taichi/ui/scene.py:17]: personally speaking only the checks in `taichi.ui.Window.__init__` is necessary because the entire GGUI is nonsense without it. And if you don't have a clear idea how to pass parameters , `string` is always a go-to solution.

**Review by PENGUINLIONG — COMMENTED** [review]
again, i still suggest you only to differentiate filled and wireframe draws, it would be a simple boolean instead of all the struggles with `DisplayMode`. I cannot come up with a case that you wanna debug meshes without the connectivity information between vertices and you still don't think `scene.particles` is a first choice. the structure of input field or ndarray ought to be the same either way.

**Comment by Morcki** [comment]
> again, i still suggest you only to differentiate filled and wireframe draws, it would be a simple boolean instead of all the struggles with `DisplayMode`. I cannot come up with a case that you wanna debug meshes without the connectivity information between vertices and you still don't think `scene.particles` is a first choice. the structure of input field or ndarray ought to be the same either way.

I agree with that, I will pass a bool value to enable users to decide to show WareFrame mode or not to.

**Review by ailzhang — APPROVED** [review]
LGTM! Thanks!
